### PR TITLE
Add FGVC-Aircraft Dataset with Hierarchical Annotations

### LINKED ISSUE #134: [Dataset] Add FGCV-Aircraft Datasets for image classification
Please add the following [Fine-Grained Visual Classification of Aircraft (FGVC-Aircraft)](https://www.robots.ox.ac.uk/~vgg/data/fgvc-aircraft/) Dataset as new _Image classification_ datasets.

As classes of Aircraft models are organized in a four-levels hierarchy, be explicit in the documentation and example for this specificity.

Please ensure consistency with previous supported dataset with
- [ ] Reuse of the existing download and cache
- [ ] Equivalent test suite
- [ ] Documentation update

### PR DESCRIPTION
**Pull request by DerrickUnleashed**
## This PR adds support for the Fine-Grained Visual Classification of Aircraft (FGVC-Aircraft) dataset as a new image classification dataset.
- Implements the FGVC-Aircraft dataset with hierarchical class annotations:
  - **Variant**
  - **Family**
  - **Manufacturer**
- Supports downloading, caching, and preprocessing using existing utilities for consistency.
- Provides clear documentation including explicit explanation of the dataset’s hierarchical structure and usage examples.
- Includes a comprehensive test suite covering all dataset splits (`train`, `test`,`trainval` and `val`) and all annotation levels.
- Includes a `default_root` method for consistent path resolution

This integration expands the collection of fine-grained image classification datasets supported, enabling users to leverage the FGVC-Aircraft dataset with ease and confidence.

Fix #134

### CODE REVIEW

**Review by cregouby — CHANGES_REQUESTED** [review]
**praise** : thanks for the contribution.
**suggestion** : Could we improve the level of end-user information on the size and time when collecting that dataset ? Documentation and/or UI message ? What is the user experience when a new user randomly pick that dataset for a try ? 
**todo** : see inline comments.

**Review by cregouby — CHANGES_REQUESTED** [review]

File: R/dataset-fgvc.R [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+#' @param transform Optional function to transform input images after loading.
+#' @param target_transform Optional function to transform labels.
+#' @param download Logical. Whether to download the dataset if not found locally. Default is `FALSE`.
+#'
+#' @return A fgvc_aircraft_dataset object representing the dataset.
+#'
+#' @examples
+#' \dontrun{
+#' root_dir <- tempfile()
+#' fgvc <- fgvc_aircraft_dataset(root = root_dir, split = "train", annotation_level = "variant", download = TRUE)
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-fgvc.R]: Agreed — I'll keep the parameter name as `root` for consistency and clarify the description if needed.

File: R/dataset-fgvc.R [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+#'
+#' Loads the fgvc Aircraft Dataset, a dataset of aircraft images with multiple annotation levels:
+#' - "variant": fine-grained aircraft variants (e.g., specific model variants)
+#' - "family": aircraft families
+#' - "manufacturer": aircraft manufacturers
+#' 
+#' The dataset supports different splits: "train", "trainval", and "test".
+#'
+#' @param root Character. Root directory for dataset storage. The dataset will be stored under `root/fgvc-aircraft-2013b`.
+#' @param split Character. Dataset split to use. One of `"train"`, `"trainval"`, or `"test"`. Default is `"train"`.
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-fgvc.R]: Yes, I’ve updated in the documentation that `trainval` is the combined training and validation set from the official dataset. I've also explained that we manually split it into `train` (80%) and `val` (20%) using a fixed seed for reproducibility.

File: R/dataset-fgvc.R [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+#' @param annotation_level Character. Level of annotation to use. One of `"variant"`, `"family"`, or `"manufacturer"`. Default is `"variant"`.
+#' @param transform Optional function to transform input images after loading.
+#' @param target_transform Optional function to transform labels.
+#' @param download Logical. Whether to download the dataset if not found locally. Default is `FALSE`.
+#'
+#' @return A fgvc_aircraft_dataset object representing the dataset.
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-fgvc.R]: Noted. I’ve updated the return documentation to explicitly state that the returned object is a `torch::dataset` with images (x) as `tensors` and labels (y) corresponding to the selected `annotation_level`.

File: R/dataset-fgvc.R [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+#' Loads the fgvc Aircraft Dataset, a dataset of aircraft images with multiple annotation levels:
+#' - "variant": fine-grained aircraft variants (e.g., specific model variants)
+#' - "family": aircraft families
+#' - "manufacturer": aircraft manufacturers
+#' 
+#' The dataset supports different splits: "train", "trainval", and "test".
+#'
+#' @param root Character. Root directory for dataset storage. The dataset will be stored under `root/fgvc-aircraft-2013b`.
+#' @param split Character. Dataset split to use. One of `"train"`, `"trainval"`, or `"test"`. Default is `"train"`.
+#' @param annotation_level Character. Level of annotation to use. One of `"variant"`, `"family"`, or `"manufacturer"`. Default is `"variant"`.
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-fgvc.R]: Thank you for the suggestion. I’ve expanded the documentation to clarify the hierarchical relationship between `variant`, `family`, and `manufacturer`, noting that not all combinations are valid. This should give users a better understanding of the structure and constraints within the dataset.

File: tests/testthat/test-dataset-fgvc.R
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply tests/testthat/test-dataset-fgvc.R:1]: I have added a test creating a `dataloader()` from `fgvc_aircraft_dataset()` and verify that each batch returns tensors with correct shapes and labels.

File: R/dataset-fgvc.R [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+             "variant" = "variants.txt",
+             "family" = "families.txt",
+             "manufacturer" = "manufacturers.txt"))
+
+    self$classes <- readr::read_lines(ann_file)
+    self$class_to_idx <- rlang::set_names(seq_along(self$classes), self$classes)
+
+    img_folder  <- fs::path(self$data_path, "data", "images")
+    labels_file <- fs::path(self$data_path, "data", 
+                            sprintf("images_%s_%s.txt", annotation_level, split))
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-fgvc.R]: Thanks for the suggestion! I've implemented the improvement by refactoring the filename construction using `glue::glue()` for better readability and consistency.

File: R/dataset-fgvc.R [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+      if (is.null(self$class_to_idx[[class_name]])) {
+        warning("Unknown class label: ", class_name)
+        next
+      }
+
+      self$image_paths[i] <- fs::path(img_folder, paste0(img_id, ".jpg"))
+      self$labels[i] <- self$class_to_idx[[class_name]]
+    }
+  },
+
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-fgvc.R]: Thanks for pointing that out! I've implemented the change by vectorizing the loop to improve performance and clarity.

File: R/dataset-fgvc.R [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+      img_id <- parts[1]
+      class_name <- trimws(paste(parts[-1], collapse = " "))
+
+
+      if (is.null(self$class_to_idx[[class_name]])) {
+        warning("Unknown class label: ", class_name)
+        next
+      }
+
+      self$image_paths[i] <- fs::path(img_folder, paste0(img_id, ".jpg"))
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-fgvc.R]: Thanks for the suggestion! I've updated the code to consistently use `glue::glue()` for all string construction to maintain clarity and uniformity throughout.

File: R/dataset-fgvc.R [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+    lines <- readLines(label_file)
+
+    if (split %in% c("train", "val")) {
+      rlang::inform(sprintf("Creating '%s' subset from 'trainval' via 80/20 split (seed = 42)", split))
+      set.seed(42)
+      idx <- sample(length(lines))
+      cutoff <- floor(0.8 * length(lines))
+      lines <- if (split == "train") lines[idx[1:cutoff]] else lines[idx[(cutoff + 1):length(lines)]]
+    }
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-fgvc.R]: Thanks for the note! I've removed the custom `80/20` split logic and now rely directly on the existing `train` and `val` splits provided by the official dataset.

File: tests/testthat/test-dataset-fgvc.R [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+  expect_true(inherits(item$x, "torch_tensor"))
+  expect_type(item$y, "integer")
+  expect_equal(as.numeric(item$y), 1)
+
+  resize_collate_fn <- function(batch) {
+    xs <- lapply(batch, function(sample) {
+      torchvision::transform_resize(sample$x, c(224, 224))
+    })
+    xs <- torch::torch_stack(xs)
+    ys <- torch::torch_tensor(sapply(batch, function(sample) sample$y), dtype = torch::torch_long())
```
Comment (cregouby) [inline tests/testthat/test-dataset-fgvc.R:24]:
```suggestion
    xs <- lapply(batch, function(item) {
      torchvision::transform_resize(item$x, c(224, 224))
    })
    xs <- torch::torch_stack(xs)
    ys <- torch::torch_tensor(sapply(batch, function(item) item$y), dtype = torch::torch_long())
```

File: R/dataset-fgvc.R [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+#' - "family": aircraft families
+#' - "manufacturer": aircraft manufacturers
+#' 
+#' The FGVC-Aircraft dataset supports the following official splits:
+#' - `"train"`: training subset
+#' - `"val"`: validation subset
+#' - `"trainval"`: combined training and validation set
+#' - `"test"`: test set with labels (used for evaluation)
+#'
+#' These splits are provided directly in the dataset archive. No manual splitting is performed.
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-fgvc.R:14]: Noted, thank you for pointing it out. I'll remove the duplicate line.

File: R/dataset-fgvc.R [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+#' and a `"family"` always belongs to exactly one `"manufacturer"`. You cannot mix or arbitrarily combine levels.
+#'
+#' Default is `"variant"`.
+#' @param transform Optional function to transform input images after loading.
+#' @param target_transform Optional function to transform labels.
+#' @param download Logical. Whether to download the dataset if not found locally. Default is `FALSE`.
+#'
+#' @return An object of class \code{fgvc_aircraft_dataset}, which behaves like a torch-style dataset.
+#' Each element is a named list with:
+#' - `x`: the image as a 3D torch tensor (C x H x W), already converted using `transform_to_tensor()`, unlike most torchvision-style datasets where images are returned as PIL objects.
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-fgvc.R]: Good point. Thank you! Yes, since varying image sizes can impact preprocessing and model design, it's definitely worth mentioning.

File: R/dataset-fgvc.R [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+#' - `"family"`: a mid-level grouping, e.g., `"Boeing 737"`, which includes multiple variants. There are 70 distinct families.
+#' - `"manufacturer"`: the coarsest level, e.g., `"Boeing"`, grouping multiple families under a single aircraft manufacturer. There are 30 manufacturers.
+#'
+#' Note: These levels form a strict hierarchy: 
+#' each `"manufacturer"` consists of multiple `"families"`, and each `"family"` contains several `"variants"`. 
+#' Not all combinations of levels are valid — for example, a `"variant"` always belongs to exactly one `"family"`, 
+#' and a `"family"` always belongs to exactly one `"manufacturer"`. You cannot mix or arbitrarily combine levels.
+#'
+#' Default is `"variant"`.
+#' @param transform Optional function to transform input images after loading.
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-fgvc.R:19]: Got it, thanks for the note. I'll update the parameter doc to clearly state the default and avoid that confusion.

File: R/dataset-fgvc.R [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+#'
+#' @examples
+#' \dontrun{
+#' fgvc <- fgvc_aircraft_dataset( split = "train", annotation_level = "variant", download = TRUE)
+#' first_item <- fgvc[1]
+#' # image tensor of first item
+#' first_item$x
+#' # label of first item
+#' first_item$y
+#' }
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-fgvc.R:48]: Thank you for this wonderful suggestion! I've removed the original example and incorporated your dataloader-based approach along with `transform_resize(size = c(768, 1024))` to better reflect the dataset's characteristics. If you'd still like the original single-sample example included alongside, I'm happy to add that back in too.

File: tests/testthat/test-dataset-fgvc.R
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply tests/testthat/test-dataset-fgvc.R:1]: Thank you so much for the improvement. Absolutely! using more expressive `expect_` helpers `expect_tensor()` , `expect_tensor_dtype()`, `expect_tensor_shape()` greatly improves clarity and debuggability of tests.

File: R/dataset-fgvc.R [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+#' One of:
+#' - `"variant"`: the most fine-grained level, e.g., `"Boeing 737-700"`. There are 100 visually distinguishable aircraft variants.
+#' - `"family"`: a mid-level grouping, e.g., `"Boeing 737"`, which includes multiple variants. There are 70 distinct families.
+#' - `"manufacturer"`: the coarsest level, e.g., `"Boeing"`, grouping multiple families under a single aircraft manufacturer. There are 30 manufacturers.
+#'
+#' Note: These levels form a strict hierarchy: 
+#' each `"manufacturer"` consists of multiple `"families"`, and each `"family"` contains several `"variants"`. 
+#' Not all combinations of levels are valid — for example, a `"variant"` always belongs to exactly one `"family"`, 
+#' and a `"family"` always belongs to exactly one `"manufacturer"`. You cannot mix or arbitrarily combine levels.
+#'
```
[the remark this answers is not in the document]
Comment (cregouby) [reply R/dataset-fgvc.R]: **improvement** the two paragraphs are now in the correct Details location, but you missed the 'merge' demand. Could you please merge those two paragraph that contains a lot of duplicate information side by side ?

File: tests/testthat/test-dataset-fgvc.R [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+  expect_type(item$y, "integer")
+  expect_equal(as.numeric(item$y), 1)
+
+  fgvc <- fgvc_aircraft_dataset(root = t, split = "trainval", annotation_level = "variant")
+  expect_length(fgvc, 6667)
+  item <- fgvc[1]
+  expect_named(item, c("x", "y"))
+  expect_tensor(item$x)
+  expect_tensor_shape(item$x,c(3,695,1024))
+  expect_tensor_dtype(item$x,torch_float())
```
Comment (DerrickUnleashed) [inline tests/testthat/test-dataset-fgvc.R:58]: Are these lines correct, or am I missing something? I'm a bit confused.

**Comment by DerrickUnleashed** [comment]
> @DerrickUnleashed, Could you please ensure that the result documentation looks good ?

Sure I have removed some more duplicacies in the documentation. It looks good now. Let me know if there are any more changes needed.

**Comment by cregouby** [comment]
> > @DerrickUnleashed, Could you please ensure that the result documentation looks good ?
> 
> Sure I have removed some more duplicacies in the documentation. It looks good now. However i did see the following warning
> 
> ```r
> Rd file 'fgvc_aircraft_dataset.Rd':
>   \examples lines wider than 100 characters:
>      dl <- torch::dataloader( dataset = fgvc, batch_size = 2, shuffle = TRUE, collate_fn = resize_collate_fn )
> These lines will be truncated in the PDF manual.
> ```
> 
> Do I need to make changes to reduce the wide lines or can we ignore it ?

Yes you should. Any Note or Warning would prevent CRAN publication and should be fixed.

File: tests/testthat/test-dataset-fgvc.R
[the remark this answers is not in the document]
Comment (cregouby) [reply tests/testthat/test-dataset-fgvc.R:1]: It is fine now, thanks !

**Review by cregouby — APPROVED** [review]
